src/java/org/apache/cassandra/cql3/statements/TxnDataName.java

```diff
@@ -0,0 +1,219 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.cql3.statements;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+import org.apache.cassandra.db.DecoratedKey;
+import org.apache.cassandra.db.TypeSizes;
+import org.apache.cassandra.io.IVersionedSerializer;
+import org.apache.cassandra.io.util.DataInputPlus;
+import org.apache.cassandra.io.util.DataOutputPlus;
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.utils.ByteBufferUtil;
+
+public class TxnDataName implements Comparable<TxnDataName>
+{
+    private static final TxnDataName RETURNING = new TxnDataName(Kind.RETURNING);
+
+    public enum Kind
+    {
+        USER((byte) 1),
+        RETURNING((byte) 2),
+        AUTO_READ((byte) 3);
+
+        private final byte value;
+
+        Kind(byte value)
+        {
+            this.value = value;
+        }
+
+        public static Kind from(byte b)
+        {
+            switch (b)
+            {
+                case 1:
+                    return USER;
+                case 2:
+                    return RETURNING;
+                case 3:
+                    return AUTO_READ;
+                default:
+                    throw new IllegalArgumentException("Unknown kind: " + b);
+            }
+        }
+    }
+
+    private final Kind kind;
+    private final String[] parts;
+
+    public TxnDataName(Kind kind, String... parts)
+    {
+        this.kind = kind;
+        this.parts = parts;
+    }
+
+    public static TxnDataName user(String name)
+    {
+        return new TxnDataName(Kind.USER, name);
+    }
+
+    public static TxnDataName returning()
+    {
+        return RETURNING;
+    }
+
+    public static TxnDataName partitionRead(TableMetadata metadata, DecoratedKey key)
+    {
+        return new TxnDataName(Kind.AUTO_READ, metadata.keyspace, metadata.name, bytesToString(key.getKey()));
+    }
+
+    private static String bytesToString(ByteBuffer bytes)
+    {
+        return ByteBufferUtil.bytesToHex(bytes);
+    }
+
+    private static ByteBuffer stringToBytes(String string)
+    {
+        return ByteBufferUtil.hexToBytes(string);
+    }
```

Comment:
I had compression logic but the more I thought about it the more I thought that was dangerous... if the input data is gzip we may attempt to unzip it and no longer find the key!